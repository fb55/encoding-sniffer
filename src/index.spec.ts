import { DecodeStream } from "./index.js";
import { readdirSync, createReadStream, promises as fs } from "node:fs";
import path from "node:path";
import { setTimeout } from "node:timers/promises";

function getStream(stream: NodeJS.ReadableStream): Promise<string> {
    // TODO[engines.node@>=18]: Use `reduce`
    return new Promise((resolve, reject) => {
        let data = "";
        stream.on("data", (chunk) => {
            expect(typeof chunk).toBe("string");
            data += chunk;
        });
        stream.on("end", () => resolve(data));
        stream.on("error", reject);
        stream.resume();
    });
}

describe("DecodeStream", () => {
    it("should decode a UTF-8 string", async () => {
        const stream = new DecodeStream();
        stream.end(Buffer.from("Hello, world!"));
        expect(await getStream(stream)).toBe("Hello, world!");
    });

    describe("Fixtures", () => {
        for (const file of readdirSync(path.join(__dirname, "__fixtures__"))) {
            if (!file.endsWith(".html")) continue;

            it(`should decode ${file}`, async () => {
                const stream = new DecodeStream();
                createReadStream(
                    path.join(__dirname, "__fixtures__", file),
                ).pipe(stream);
                expect(await getStream(stream)).toMatchSnapshot();
            });
        }
    });

    it("should decode a file one byte at a time", async () => {
        const file = await fs.readFile(
            path.join(__dirname, "__fixtures__", "utf-16be-bom.html"),
        );
        const stream = new DecodeStream();
        const collector = getStream(stream);
        for (let index = 0; index < file.length; index++) {
            // Wait for a bit to allow the stream to process the data.
            await setTimeout(0);
            stream.write(file.slice(index, index + 1));
        }
        stream.end();
        expect(await collector).toMatchSnapshot();
    });
});
