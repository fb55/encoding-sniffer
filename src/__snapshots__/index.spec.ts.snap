// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DecodeStream > Fixtures > should decode no-bom-charset-bracket.html 1`] = `
"<!DOCTYPE html>
<html>
    <head><meta charset="utf-8">
    </head>
    <body></body>
</html>
"
`;

exports[`DecodeStream > Fixtures > should decode no-bom-charset-http-equiv-no-quotes.html 1`] = `
"<!DOCTYPE html>
<html>
  <head>
      <!--<meta http-equiv="Content-Type" content="text/html;charset=ISO-8859-8">-->
      <meta http-equiv=Content-Type content=text/html;charset=ISO-8859-5>
      <script src="/resources/testharness.js"></script>
      <script src="/resources/testharnessreport.js"></script>
  </head>
  <body></body>
</html>
"
`;

exports[`DecodeStream > Fixtures > should decode no-bom-charset-http-equiv-refresh.html 1`] = `
"<!DOCTYPE html>
<html>
    <head>
        <meta http-equiv=refresh http-equiv=Content-Type content="text/html;charset=iso-8859-2">
    </head>
    <body></body>
</html>
"
`;

exports[`DecodeStream > Fixtures > should decode no-bom-charset-http-equiv-second-charset.html 1`] = `
"<!DOCTYPE html>
<html>
    <head>
        <meta http-equiv=Content-Type content="text/html;charsetcharset=iso-8859-2">
    </head>
    <body></body>
</html>
"
`;

exports[`DecodeStream > Fixtures > should decode no-bom-charset-http-equiv-tis-620.html 1`] = `
"<!DOCTYPE html>
<html>
    <head>
        <script src="script-utf8.js" charset="UTF-8"></script>
        <script src="script-iso88591.js" charset="ISO-8859-1"></script>
        <meta http-equiv="content-type" content="text/html;charset='tis-620'">
    </head>
    <body>This is not really tis-620 (windows-874), of course</body>
</html>
"
`;

exports[`DecodeStream > Fixtures > should decode no-bom-charset-http-equiv-trailing-space.html 1`] = `
"<!DOCTYPE html>
<html>
    <head>
        <meta http-equiv=Content-Type content="text/html;charset=iso-8859-2 ">
    </head>
    <body></body>
</html>
"
`;

exports[`DecodeStream > Fixtures > should decode no-bom-charset-koi8.html 1`] = `
"<!DOCTYPE html>
<html>
    <head>
        <script src="script-utf8.js" charset="UTF-8"></script>
        <script src="script-iso88591.js" charset="ISO-8859-1"></script>
        <meta charset="koi8">
    </head>
    <body>This is not really koi8 (KOI8-R), of course</body>
</html>
"
`;

exports[`DecodeStream > Fixtures > should decode no-bom-charset-short-comment.html 1`] = `
"<!DOCTYPE html>
<html>
    <head>
    <!--><meta charset="iso-8859-2">
    </head>
    <body></body>
</html>
"
`;

exports[`DecodeStream > Fixtures > should decode no-bom-charset-utf-16.html 1`] = `
"<!DOCTYPE html>
<html>
<meta charset="utf-16">
"
`;

exports[`DecodeStream > Fixtures > should decode no-bom-charset-utf-16be.html 1`] = `
"<!DOCTYPE html>
<html>
<meta charset="utf-16be">
"
`;

exports[`DecodeStream > Fixtures > should decode no-bom-charset-utf-16le.html 1`] = `
"<!DOCTYPE html>
<html>
<meta charset="utf-16le">
"
`;

exports[`DecodeStream > Fixtures > should decode no-bom-charset-x-user-defined.html 1`] = `
"<!DOCTYPE html>
<html>
<meta charset="x-user-defined">
"
`;

exports[`DecodeStream > Fixtures > should decode no-bom-no-charset.html 1`] = `
"<!DOCTYPE html>
<html>
    <head>
        <script src="script-utf8.js" charset="UTF-8"></script>
        <script src="script-iso88591.js" charset="ISO-8859-1"></script>
    </head>
    <body>©</body>
</html>"
`;

exports[`DecodeStream > Fixtures > should decode utf-8-bom.html 1`] = `
"<!DOCTYPE html>
<html>
    <head>
        <meta charset="ISO-8859-4"></script>
        <script src="script-utf8.js" charset="UTF-8"></script>
        <script src="script-iso88591.js" charset="ISO-8859-1"></script>
    </head>
    <body>©</body>
</html>"
`;

exports[`DecodeStream > Fixtures > should decode utf-16be-bom.html 1`] = `
"<!DOCTYPE html>
<html>
    <head>
        <meta charset="ISO-8859-4"></script>
        <script src="script-utf8.js" charset="UTF-8"></script>
        <script src="script-iso88591.js" charset="ISO-8859-1"></script>
    </head>
    <body>©</body>
</html>"
`;

exports[`DecodeStream > Fixtures > should decode utf-16le-bom.html 1`] = `
"<!DOCTYPE html>
<html>
    <head>
        <meta charset="ISO-8859-4"></script>
        <script src="script-utf8.js" charset="UTF-8"></script>
        <script src="script-iso88591.js" charset="ISO-8859-1"></script>
    </head>
    <body>©</body>
</html>"
`;

exports[`DecodeStream > should decode a file one byte at a time 1`] = `
"<!DOCTYPE html>
<html>
    <head>
        <meta charset="ISO-8859-4"></script>
        <script src="script-utf8.js" charset="UTF-8"></script>
        <script src="script-iso88591.js" charset="ISO-8859-1"></script>
    </head>
    <body>©</body>
</html>"
`;
